### ISSUE TITLE
JX_MakePersistent() documentation

### ISSUE THREAD

**Issue by treasurebox** [jxcore/jxcore#617]
Hi!

  I find JX_MakePersistent() very-very useful, however I find that it is not documented well enough. I feel that to use it properly, I would like to understand it better. A few questions that came to my mind:

- If I get a JXValue argument in an extension, and call JX_MakePersistent() to hold on to it after the function returns, then I would need to JX_Free() it myself when I am done to it, right? But I don't need JX_ClearPersistent(), just JX_Free() is enough?
- If I don't really need to hold into said argument, but nevertheless I call JX_MakePersistent() and JX_Free() every time (suppose I want to write a higher level abstraction), does it incur any performance or memory penalty? And if yes, how bad is it?
- The other direction: If I create a new object on client side, and pass it back as result value (args+argc), JX_Free() -ing it would lead to crashing. But if I have done a JX_MakePersistent() on given object, is it ok, or I really shouldn't do that?
- If I just create a new object on client side, and don't really pass it back to javascript side then, but use later, should I call JX_MakePersistent() on it before returning from extension?

Thanks!

**Comment by obastemur** [comment]
`JX_Free` is to free the object that you have created. However, `JX_ClearPersistent` needs to be called for an object that you've marked as `Persistent`. So, you should call `JX_ClearPersistent`

You can definitely use C++ features to make things more automatic. However, this API needs to be C. Don't think there will be a big perf penalty though.

Does it really crash ? Post a test code pls.

You should make persistent anything on the native side that doesn't have any reference. Otherwise, that object will be GC'ed

**Comment by treasurebox** [comment]
No crash, just want to understand it better :)

So, if I get an object as argument, and I want to hold to it, I call JX_MakePersistent(), then later when I free it, I need JX_ClearPersistent(), but must not call JX_Free() ? (as it is not an object I've created myself)

**Comment by obastemur** [comment]
Better call `JX_Free` anytime you are done with an object. It won't be GC'ing real JS object anyway. That's the job for JS engine. Instead it will cleanup the native wrapper etc. 

So, call JS_MakePersistent + JS_Free

**Comment by treasurebox** [comment]
So no need for ClearPersistent() after all if I call Free(), thanks!

**Comment by obastemur** [comment]
no, if you don't call clear persistent, it's not going to be freed by JS GC!

**Comment by treasurebox** [comment]
Ah, now I get the point! So it effectively turns gc off/on for given object.

**Comment by obastemur** [comment]
@treasurebox Yes indeed. Actually if anyone can contribute documenting this, I would hugely appreciate!
